internal/webhook/v1alpha1/testrun_webhook.go

```diff
@@ -0,0 +1,212 @@
+// Copyright Axis Communications AB.
+//
+// For a full list of individual contributors, please see the commit history.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//	http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package v1alpha1
+
+import (
+	"context"
+	"fmt"
+
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	"k8s.io/apimachinery/pkg/util/uuid"
+	"k8s.io/apimachinery/pkg/util/validation/field"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	logf "sigs.k8s.io/controller-runtime/pkg/log"
+	"sigs.k8s.io/controller-runtime/pkg/webhook"
+	"sigs.k8s.io/controller-runtime/pkg/webhook/admission"
+
+	etosv1alpha1 "github.com/eiffel-community/etos/api/v1alpha1"
+)
+
+// nolint:unused
+// log is for logging in this package.
+var testrunlog = logf.Log.WithName("testrun-resource")
+
+// SetupTestRunWebhookWithManager registers the webhook for TestRun in the manager.
+func SetupTestRunWebhookWithManager(mgr ctrl.Manager) error {
+	if cli == nil {
+		cli = mgr.GetClient()
+	}
+	return ctrl.NewWebhookManagedBy(mgr).For(&etosv1alpha1.TestRun{}).
+		WithValidator(&TestRunCustomValidator{}).
+		WithDefaulter(&TestRunCustomDefaulter{}).
+		Complete()
+}
+
+// +kubebuilder:webhook:path=/mutate-etos-eiffel-community-github-io-v1alpha1-testrun,mutating=true,failurePolicy=fail,sideEffects=None,groups=etos.eiffel-community.github.io,resources=testruns,verbs=create;update,versions=v1alpha1,name=mtestrun-v1alpha1.kb.io,admissionReviewVersions=v1
+
+// TestRunCustomDefaulter struct is responsible for setting default values on the custom resource of the
+// Kind TestRun when those are created or updated.
+//
+// NOTE: The +kubebuilder:object:generate=false marker prevents controller-gen from generating DeepCopy methods,
+// as it is used only for temporary operations and does not need to be deeply copied.
+type TestRunCustomDefaulter struct{}
+
+var _ webhook.CustomDefaulter = &TestRunCustomDefaulter{}
+
+// Default implements webhook.CustomDefaulter so a webhook will be registered for the Kind TestRun.
+func (d *TestRunCustomDefaulter) Default(ctx context.Context, obj runtime.Object) error {
+	testrun, ok := obj.(*etosv1alpha1.TestRun)
+
+	if !ok {
+		return fmt.Errorf("expected an TestRun object but got %T", obj)
+	}
+	testrunlog.Info("Defaulting for TestRun", "name", testrun.GetName())
```

Comment:
Assuming that we're propagating/setting default values here, then we could also spell that out more clearly in the log. E.g Setting default value for testrun or similar.